### PR TITLE
Prevent pattern matches across mutation ops in inductor pre-grad FX passes

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101144

Per https://github.com/pytorch/pytorch/issues/101124

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/pattern_matcher.py [hunk 1]
```diff
@@ -711,6 +711,36 @@ def decorator(handler):
     return decorator
 
 
+def is_start_of_fx_graph(node):
+    return len(node.all_input_nodes) == 0
```
Comment (jansel) [inline torch/_inductor/pattern_matcher.py:715]: This doesn't seem right.  Wouldn't this be true for `torch.tensor(0)` or `torch.zeros([])`?

**Review by jansel — CHANGES_REQUESTED** [review]

**Comment by jansel** [comment]
> Running `pytest test/inductor` now takes significantly longer. Runtime increased to 10788.64s from 1691.55s.

That is weird.  You could disable this for joint_graph and post_grad, and only turn it on for the pre-grad graph.

**Comment by williamwen42** [comment]
Is there an existing method to do this, or do I just add the mutation check directly to the existing passes in pre_grad.py?

File: torch/_inductor/pattern_matcher.py (hunk 1 above)
Comment (williamwen42) [reply torch/_inductor/pattern_matcher.py:715]: I think this depends on what we mean by the "start of the FX graph". I implemented it as "a node that has no inputs," but perhaps there needs to be the additional condition "a node that is before any mutation op"?

Comment (jansel) [reply torch/_inductor/pattern_matcher.py:715]: It should be the literal first node in the graph.

**Comment by jansel** [comment]
> Is there an existing method to do this, or do I just add the mutation check directly to the existing passes in pre_grad.py?

You would need to add a flag (or a mode).

**Comment by williamwen42** [comment]
Inductor unittests don't seem to be taking substantially longer anymore.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**